Move regression tests that use C library functions to a separate folder

tautschnig opened the pull request:
This is a first step to disentangle testing of CBMC's algorithms and core
components from testing of the library models. The next step is to ensure
coverage of all modelled library functions.

Marking this do-not-merge as there are further tests to be moved, and some tests to be split into multiple, smaller tests. Nevertheless, early feedback as to whether this is a good idea at all would be much appreciated.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Move regression tests that use C library functions to a separate folder

This is a first step to disentangle testing of CBMC's algorithms and core
components from testing of the library models. The next step is to ensure
coverage of all modelled library functions.

tautschnig committed:
clang-format the moved regression tests

This commit has only whitespace changes.

martin-cs commented:
> Nevertheless, early feedback as to whether this is a good idea at all would be much appreciated.

[+] This fits better with the development model that seems to be happening where library models are increasingly independent and their own repositories.
[+] Clarity about what is being tested is always a good thing.
[-] How intrusive are this patch set likely to be?

tautschnig commented:
I'd now consider this ready for review. In a subsequent step I'll propose new regression tests for library functions currently not being exercised.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 254d09782f3c893ecd6044e2d2709a8fc28d00df).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89812749

martin-cs commented:
Views mostly as above but a PR this size should really get buy-in from multiple people.  @chrisr-diffblue @smowton @peterschrammel what do you think?

smowton commented:
The idea seems sensible to me, but I can't see myself having time to provide a meaningful review

chrisr-diffblue commented on regression/cbmc-library/CreateThread1/main.c:
```diff
@@ -0,0 +1,9 @@
+#include <windows.h>
```
Minor nit pick - could we slightly tweak the directory naming for these test? having a number postfix with no obvious separator gets a bit confusing where the function being tested *also* has a numeric postfix (e.g. `_InterlockedAnd161` is the first test of  the `_InterlockedAnd16` function) - perhaps a naming scheme like `functionUnderTest-N` might be clearer? Or maybe even sub-directories, but thats probably more mechanics than we'd want to get into.

tautschnig replied:
Indeed, that was very confusing. I have now used a `-01` suffix instead of `1`.

chrisr-diffblue commented:
I'm really pleased to see some cleanup to our tests, which are quite a mess to navigate - so I'm definitely supportive. One minor nit pick, but thats all. (Thanks @martin-cs for flagging up this PR)

tautschnig committed:
Add KNOWNBUG regression test skeletons for all modelled C/C++ library functions

We model a number of library functions, but do not necessarily verify them
against any kind of specification. We should make sure that they are at least
free from memory errors. This commit introduces them effectively as a long to-do
list, where each test should eventually be completed and marked as CORE.

The tests were generated using the following script, with a manual removal of
java::java.io.InputStream.read:()I-01 afterwards:

 #!/bin/bash

 for lib in src/ansi-c/library/*.c src/cpp/library/*.c
 do
   echo "lib: $lib"
   header=$(basename $lib .c)
   echo "header: $header"

   for f in $(git grep FUNCTION: $lib | cut -f3 -d" " | sort | uniq)
   do
     echo "f: $f"
     test_dir="regression/cbmc-library/${f}-01"
     if [ ! -d $test_dir ]
     then
       mkdir -p $test_dir

       cat > $test_dir/main.c <<EOF
 #include <$header.h>
 #include <assert.h>

 int main()
 {
   $f();
   assert(0);
   return 0;
 }
 EOF
       git add $test_dir/main.c

       cat > $test_dir/test.desc <<EOF
 KNOWNBUG
 main.c
 --pointer-check --bounds-check
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
 ^warning: ignoring
 EOF
       git add $test_dir/test.desc
     fi
   done
 done

bla

tautschnig commented:
As reviewing the last commit in this PR is obviously somewhat pointless, here is the script that I've used to generate the tests:
````bash
#!/bin/bash

for lib in src/ansi-c/library/*.c src/cpp/library/*.c
do
  echo "lib: $lib"
  header=$(basename $lib .c)
  echo "header: $header"

  for f in $(git grep FUNCTION: $lib | cut -f3 -d" " | sort | uniq)
  do
    echo "f: $f"
    test_dir="regression/cbmc-library/${f}-01"
    if [ ! -d $test_dir ]
    then
      mkdir -p $test_dir

      cat > $test_dir/main.c <<EOF
#include <$header.h>
#include <assert.h>

int main()
{
  $f();
  assert(0);
  return 0;
}
EOF
      git add $test_dir/main.c

      cat > $test_dir/test.desc <<EOF
KNOWNBUG
main.c
--pointer-check --bounds-check
^EXIT=0$
^SIGNAL=0$
^VERIFICATION SUCCESSFUL$
--
^warning: ignoring
EOF
      git add $test_dir/test.desc
    fi
  done
done
````

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6d852fb62145639c8d3bf24a2394f7ce2020369a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90469383
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 30ce180f14d49eb54f4c3abca9a0d3897da94ab2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90470682
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e7330a04f9d195672e702d89c264106aaedff017).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90472111

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f37b54bfbb778061f43bc843dfdb70b08387a984).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90492359

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5ec635030011c627bd3756082f482b2f7b07892d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90496872
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e8b00cd18ec1cb2bbf8b839798002ea391014bd9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90498444

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2b0ad13e6cd9c5683d8e365b1517da45111c5fb5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90525242

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 3534b823b8fc74c98eb5c85233211728d45c05b2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90583796
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 122c57cd1cd120082d6780b63533c48b9209d597).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90586487
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 61062cb8ed7d187b03c86833b5303dbc2e07d39a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90588740
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: dfea0a69a3bd69fb38841fffeecec6c377c46fab).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90587440
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a4d2828ffafe17801c9807ca301404ee399bcf4f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90854570
